OPDATA-3708 USDO add Solana reserves to token balance

Subarna-Singh opened the pull request:
## Closes #OPDATA-3708
https://smartcontract-it.atlassian.net/browse/OPDATA-3708

## Description
USDO PoR is being extended to include Solana-based reserves, beginning with TBILL tokens that back USDO on Solana.
The token-balance EA should:

- Query the TBILL (SOL) token account at 4MmJVdwYN8LwvbGeCowYjSx7KoEi6BJWg8XXnW4fDDp6 to fetch the balance.
- Retrieve the TBILL price from the on-chain price oracle (Ethereum or Arbitrum).
- Multiply the balance by the oracle price to calculate the USD value of Solana reserves.
- Contribute this USD value to the composite USDO PoR output so it reflects Solana reserves alongside existing assets.

## Changes
Added new endpoint usdoSolana.

## Steps to Test
yarn test packes/source/token-balance/test/unit
yarn test packes/source/token-balance/test/integration

## Quality Assurance

- [ ] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `infra-k8s` configuration file.
- [ ] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `adapter-secrets` configuration file or update the [soak testing blacklist](/packages/scripts/src/get-changed-adapters/soakTestBlacklist.ts).
- [ ] If a new adapter was made, or a new endpoint was added, update the `test-payload.json` file with relevant requests.
- [ ] The branch naming follows git flow (`feature/x`, `chore/x`, `release/x`, `hotfix/x`, `fix/x`) or is created from Jira.
- [x] This is related to a maximum of one Jira story or GitHub issue.
- [x] Types are safe (avoid TypeScript/TSLint features like any and disable, instead use more specific types).
- [x] All code changes have 100% unit and integration test coverage. If testing is not applicable or too difficult to justify doing, the reasoning should be documented explicitly in the PR.

dskloetc commented on packages/sources/token-balance/src/endpoint/usdoSolana.ts:
```diff
@@ -0,0 +1,102 @@
+      },
+      priceOracle: {
+        contractAddress: '0xCe9a6626Eb99eaeA829D7fA613d5D0A2eaE45F40',
+        chainId: '1',
```
Can this `chainId` be different from the one on line 71?
It has to match for us to be able to look up the price, right?

dskloetc commented on packages/sources/token-balance/src/transport/solana.ts:
```diff
@@ -16,7 +16,7 @@ export const getToken = async (
   const response = await Promise.all(
     addresses
-      .filter((a) => a.token?.toLowerCase() == token)
+      .filter((a) => a.token?.toLowerCase() == token.toLowerCase())
```
Is this related to this PR?
If so, can you cover it in a unit test?

Subarna-Singh replied:
Removed it from this PR.

Subarna-Singh commented on packages/sources/token-balance/src/endpoint/usdoSolana.ts:
```diff
@@ -0,0 +1,102 @@
+    },
+    tokenMint: {
+      required: true,
+      description: 'token mint',
```
added.

Subarna-Singh commented on packages/sources/token-balance/src/endpoint/usdoSolana.ts:
```diff
@@ -0,0 +1,102 @@
+        contractAddress: {
+          required: true,
+          type: 'string',
+          description: 'Address of token mint address',
```
redundant

Subarna-Singh commented on packages/sources/token-balance/src/transport/utils.ts:
```diff
@@ -99,3 +101,49 @@ export const getNetworkEnvVar = (network: string, suffix: string): string => {
   return envVar
 }
+
+export const getTokenBalance = async (
```
I looked into this. However, it expects address to have wallets which does not fit inti the address structure in this case. 

I will make some changes to make the function modular to allow using parts of it in future.

dskloetc replied:
> I will not be making any changes to getToken in this PR.

Why not? If the logic is the same but only the input type is different, can we just have a wrapper function that does the conversion instead of have 2 full implementations of the same logic?

dskloetc commented on packages/sources/token-balance/src/transport/usdoSolana.ts:
```diff
@@ -0,0 +1,146 @@
+
+export class USDOSolanaTransport extends SubscriptionTransport<BaseEndpointTypes> {
+  ethProvider!: ethers.JsonRpcProvider
+  arbProvider!: ethers.JsonRpcProvider
```
Is this unused?

dskloetc commented on packages/sources/token-balance/src/transport/usdoSolana.ts:
```diff
@@ -0,0 +1,146 @@
+    const tokenMint = param.tokenMint
+    const providerDataRequestedUnixMs = Date.now()
+
+    // 1. Fetch TBILL price ONCE from oracle contract
```
The logic has abstracted away whether this is TBILL, right?
Just stating in the comment that it's TBILL is confusing if it can be any token.

dskloetc commented on packages/sources/token-balance/src/transport/usdoSolana.ts:
```diff
@@ -0,0 +1,146 @@
+    const providerDataRequestedUnixMs = Date.now()
+
+    // 1. Fetch TBILL price ONCE from oracle contract
+    const priceFeed = await getTokenPrice({
```
What are you trying to indicate by putting "feed" in the name?

dskloetc commented on packages/sources/token-balance/src/transport/usdoSolana.ts:
```diff
@@ -0,0 +1,146 @@
+    tokenMint: typeof inputParameters.validated.tokenMint,
+    priceFeed: { value: bigint; decimal: number },
+  ): Promise<bigint> {
+    //1. Fetch TBILL token balances for the given address on Solana
```
Again "TBILL".

dskloetc commented on packages/sources/token-balance/src/transport/usdoSolana.ts:
```diff
@@ -0,0 +1,146 @@
+    const { result: balances } = await getTokenBalance(addresses, tokenMint, this.connection)
+
+    // 2. Multiply Solana token balances by oracle price and normalize decimals to RESULT_DECIMALS
+    let totalSharesUSD = BigInt(0)
```
What is "shares" indicating here. We don't typically speak of USD amounts in terms of "shares", right?

dskloetc commented on packages/sources/token-balance/src/transport/usdoSolana.ts:
```diff
@@ -0,0 +1,146 @@
+    let totalSharesUSD = BigInt(0)
+    for (const bal of balances) {
+      // Normalize token balance decimals to RESULT_DECIMALS
+      const normalizedShares = bal.value * BigInt(10 ** (RESULT_DECIMALS - bal.decimals))
```
Does this work if `bal-decimals > RESULT_DECIMALS`?
I think it's safer to multiply and divide: `bal.value * 10n ** BigInt(RESULT_DECIMALS) / 10n ** BitInt(bal.decimals)`

But I think it's even safer to first add all the values together and then do the decimal conversion. And this should be fine since each balance should have the same decimals, right?

dskloetc commented on packages/sources/token-balance/src/transport/solana.ts:
```diff
@@ -51,3 +65,82 @@ export const getToken = async (
+  addresses: address[],
+  tokenMint: tokenMint,
+  connection?: Connection,
+) => {
```
Can you declare a return type explicitly?

dskloetc commented on packages/sources/token-balance/src/transport/solana.ts:
```diff
@@ -51,3 +65,82 @@ export const getToken = async (
+function extractBalances(
+  response: Awaited<ReturnType<typeof fetchTokenAccounts>>,
+  cachedDecimals: number,
+) {
```
Can you declare a return type explicitly?

dskloetc commented on packages/sources/token-balance/src/transport/solana.ts:
```diff
@@ -51,3 +65,82 @@ export const getToken = async (
+    token: v.account.data.parsed.info.mint,
+    wallet: v.account.data.parsed.info.owner,
+    value: BigInt(v.account.data.parsed.info.tokenAmount.amount),
+    decimals: Number(v.account.data.parsed.info.tokenAmount.decimals ?? cachedDecimals),
```
Can `v.account.data.parsed.info.tokenAmount.decimals` be present and different from `cachedDecimals`?
Can `v.account.data.parsed.info.tokenAmount.decimals` be absent?

Based on [this documentation](https://solana.com/docs/rpc/json-structures#token-balances)
> decimals: <number> - Number of decimals configured for token's mint.

it's always equal to the decimals for the token.

dskloetc commented on packages/sources/token-balance/src/transport/solana.ts:
```diff
@@ -51,3 +65,82 @@ export const getToken = async (
+
+  return response.value.map((v) => ({
+    token: v.account.data.parsed.info.mint,
+    wallet: v.account.data.parsed.info.owner,
```
Why include `token` and `wallet` while they are unused?

dskloetc commented on packages/sources/token-balance/src/transport/usdoSolana.ts:
```diff
@@ -0,0 +1,146 @@
+      // Multiply balance by oracle price and normalize again
+      totalSharesUSD +=
+        (normalizedShares * priceFeed.value * BigInt(10 ** (RESULT_DECIMALS - priceFeed.decimal))) /
+        BigInt(10 ** RESULT_DECIMALS)
```
You're both multiplying and dividing by `10 ** RESULT_DECIMALS` here. Doesn't that cancel out.

Subarna-Singh commented on packages/sources/token-balance/src/endpoint/index.ts:
```diff
@@ -2,4 +2,5 @@ export { endpoint as etherFi } from './etherFi'
 export { endpoint as evm } from './evm'
 export { endpoint as solvJlp } from './solvJlp'
 export { endpoint as tbill } from './tbill'
+export { endpoint as usdoSolana } from './usdoSolana'
```
A generic name is a good idea.

Subarna-Singh replied:
Using solanaTokenBalance as new name

dskloetc replied:
token-balance is already the name of the EA. And other endpoints also don't have "TokenBalance" in the name.
Is there something specific about solana that the endpoint needs to be distinguished as a "token balance" endpoint in particular?

One thing that might be worth distinguishing (although I failed to do that with `xrpl`) is that the result is converted with a price contract. Or is that done for all endpoints aready?

For now I think the endpoint could just be called `solana`?

dskloetc commented on packages/sources/token-balance/src/transport/solana.ts:
```diff
@@ -0,0 +1,148 @@
+
+    if (!adapterSettings.SOLANA_RPC_URL) {
+      logger.error('SOLANA_RPC_URL is missing')
+      throw new AdapterInputError({
```
Sorry, this is my mistake!
After seeing the need to change the existing integration tests, I realized that not all NOPs that run `token-balance` might need to be able to handle the solana endpoint. Only those that get a job that needs it.
So we should definitely remove the error and log only a warning. Not an error.

Otherwise this would definitely be a major breaking change requiring all NOPs to support solana immediately.

dskloetc commented on packages/sources/token-balance/test/integration/solana.test.ts:
```diff
@@ -0,0 +1,135 @@
+              data: {
+                parsed: {
+                  info: {
+                    mint: '4MmJVdwYN8LwvbGeCowYjSx7KoEi6BJWg8XXnW4fDDp6',
```
Use (and reuse) a constant for the mint address.

dskloetc commented on packages/sources/token-balance/test/integration/solana.test.ts:
```diff
@@ -0,0 +1,135 @@
+                parsed: {
+                  info: {
+                    mint: '4MmJVdwYN8LwvbGeCowYjSx7KoEi6BJWg8XXnW4fDDp6',
+                    owner: 'G7v3P9yPtBj1e3JN7B6dq4zbkrrW3e2ovdwAkSTKuUFG',
```
Use a constant.

dskloetc commented on packages/sources/token-balance/test/integration/solana.test.ts:
```diff
@@ -0,0 +1,135 @@
+                    mint: '4MmJVdwYN8LwvbGeCowYjSx7KoEi6BJWg8XXnW4fDDp6',
+                    owner: 'G7v3P9yPtBj1e3JN7B6dq4zbkrrW3e2ovdwAkSTKuUFG',
+                    tokenAmount: {
+                      amount: '1000000000', // 1e9
```
You can write `String(1e9)`. Then you don't need the comment and you don't need to trust that the comment is correct.

dskloetc commented on packages/sources/token-balance/test/integration/solana.test.ts:
```diff
@@ -0,0 +1,135 @@
+                    owner: 'G7v3P9yPtBj1e3JN7B6dq4zbkrrW3e2ovdwAkSTKuUFG',
+                    tokenAmount: {
+                      amount: '1000000000', // 1e9
+                      decimals: 6,
```
Use a constant.

dskloetc commented on packages/sources/token-balance/test/integration/solana.test.ts:
```diff
@@ -0,0 +1,135 @@
+    async getBalance() {
+      return 0
+    }
+    async getParsedAccountInfo() {
```
Is this used? Do we need it?
If yes, does it need to return the same values as above?
If yes, use the same constants.
If no, use different values.

dskloetc commented on packages/sources/token-balance/test/integration/solana.test.ts:
```diff
@@ -0,0 +1,135 @@
+    ethers: {
+      JsonRpcProvider: jest.fn().mockImplementation(() => ({} as JsonRpcProvider)),
+      Contract: jest.fn().mockImplementation(() => ({
+        decimals: jest.fn().mockResolvedValue(8),
```
Use a constant.

dskloetc commented on packages/sources/token-balance/test/integration/solana.test.ts:
```diff
@@ -0,0 +1,135 @@
+      JsonRpcProvider: jest.fn().mockImplementation(() => ({} as JsonRpcProvider)),
+      Contract: jest.fn().mockImplementation(() => ({
+        decimals: jest.fn().mockResolvedValue(8),
+        latestAnswer: jest.fn().mockResolvedValue(150000000n), // 1.5 USD
```
Use a constant.
You can declare it as
```
const tokenPriceValue = BigInt(1.5 * 10 ** tokenPriceDecimals)
```
so the value is then clear without a need to comment.

dskloetc commented on packages/sources/token-balance/test/integration/solana.test.ts:
```diff
@@ -0,0 +1,135 @@
+      Contract: jest.fn().mockImplementation(() => ({
+        decimals: jest.fn().mockResolvedValue(8),
+        latestAnswer: jest.fn().mockResolvedValue(150000000n), // 1.5 USD
+        latestRoundData: jest.fn().mockResolvedValue({
```
Do we need both `latestRoundData` and `latestAnswer`? Are both used?
If yes, do they need to return the same values?
If yes, reuse the same constants.
If no, return different values.

dskloetc commented on packages/sources/token-balance/test/unit/solana.test.ts:
Please use constants in this file as well.

dskloetc commented on packages/sources/token-balance/test/unit/solana.test.ts:
```diff
@@ -0,0 +1,131 @@
+  })
+
+  it('test scaling of calculates correct USD result', async () => {
+    jest.mocked(getTokenPrice).mockResolvedValue({ value: 900000000n, decimal: 8 })
```
FYI, you can write numbers like this to make them easier to read: `900_000_000n`

dskloetc commented on packages/sources/token-balance/test/unit/solana.test.ts:
```diff
@@ -0,0 +1,131 @@
+    )
+
+    expect(resp.statusCode).toBe(200)
+    expect(resp.result).toBe('90000000000000000000')
```
It's not immediately obvious that this is the correct result.
If you declare some values like this at the top, it's clearer:
```
const tokenPrice = 9.0;
const tokenPriceDecimals = 8;
const tokenBalance = 1.0;
const tokenBalanceDecimals = 6;
const expectedUsdBalance = tokenBalance * tokenPrice;
jest.mocked(getTokenPrice).mockResolvedValue({ value: BigInt(tokenPrice * 10 ** tokenPriceDecimals), decimal: tokenPriceDecimals })
jest.mocked(getTokenBalance).mockResolvedValue({
  result: [{ value: BigInt(tokenBalance * 10 ** tokenBalanceDecimals), decimals: tokenBalanceDecimals }],
})
```
and then here
```
expect(resp.result).toBe(String(expectedUsdBalance * 10 ** RESULT_DECIMALS))
```

dskloetc commented on packages/sources/token-balance/test/unit/solana.test.ts:
```diff
@@ -0,0 +1,131 @@
+      result: '90000000000000000000',
+      decimals: 18,
+    })
+  })
```
Please also add a test with multiple balances that need to be added together.

dskloetc commented on packages/sources/token-balance/test/unit/solana.test.ts:
```diff
@@ -0,0 +1,131 @@
+        },
+      }),
+    ).rejects.toThrow('balance fail')
+  })
```
Also add a test with different balances with different decimals which should cause an error.

Subarna-Singh commented:
@dskloetc I have fixed the issues based on your comment. 
Including 

1. using getToken [will fix the function in another PR, as you suggested]
2. Added error handling, if getParsedTokenAccountsByOwner returns no value. 

Cases where value will be empty ([]):

- Wallet never received that token
- Token account closed
- Wrong mint PublicKey
- Wrong wallet PublicKey
- RPC node lag / data not yet finalized (rare)

Subarna-Singh commented on packages/sources/token-balance/src/transport/solana.ts:
```diff
@@ -0,0 +1,158 @@
+    param: RequestParams,
+  ): Promise<AdapterResponse<BaseEndpointTypes['Response']>> {
+    const addresses = param.addresses
+    const tokenMint = param.tokenMint
```
cleaner for sure

Subarna-Singh commented on packages/sources/token-balance/src/transport/solana-utils.ts:
```diff
@@ -30,12 +30,18 @@ export const getToken = async (
-    .flatMap((r) => r.value)
-    .map((v) => ({
+  const result = response.flatMap((r, i) => {
+    if (!r.value || r.value.length === 0) {
```
I was thinking about what would break this particular piece of code. Searched about solana accounts and found out that it would not return any value in the cases list earlier by me. 

I will open a separate PR for this to isolate the change incase it breaks existing feeds.

Subarna-Singh replied:
It is anticipated. 
I'll look into existing feeds.

Subarna-Singh commented on packages/sources/token-balance/test/unit/solana.test.ts:
```diff
@@ -20,112 +26,283 @@ describe('solanaTransport._handleRequest', () => {
+  const expectedUsdBalance = tokenBalanceValue * tokenPriceValue
+  jest
+    .mocked(getTokenPrice)
+    .mockResolvedValue({
```
Addressed/
